content/named-tuples.md

```diff
@@ -0,0 +1,777 @@
+Similar to `Tuple`, the `NamedTuple` object also contains types such as `Elem`, `Head`, `Concat`
+that describe the results of these extension methods.
+
+The translation of named tuples to instances of `NamedTuple` is fixed by the specification and therefore known to the programmer. This means that:
+
+ - All tuple operations also work with named tuples "out of the box".
+ - Macro libraries can rely on this expansion.
+
+### Computed Field Names
+
+The `Selectable` trait now has a `Fields` type member that can be instantiated
+to a named tuple.
+
+```scala
+trait Selectable:
+  type Fields <: NamedTuple.AnyNamedTuple
+```
+
+If `Fields` is instantiated in a subclass of `Selectable` to some named tuple type,
+then the available fields and their types will be defined by that type. Assume `n: T`
+is an element of the `Fields` type in some class `C` that implements `Selectable`,
+that `c: C`, and that `n` is not otherwise legal as a name of a selection on `c`.
+Then `c.n` is a legal selection, which expands to `c.selectDynamic("n").asInstanceOf[T]`.
+
+It is the task of the implementation of `selectDynamic` in `C` to ensure that its
+computed result conforms to the predicted type `T`
+
+As an example, assume we have a query type `Q[T]` defined as follows:
+
+```scala
+trait Q[T] extends Selectable:
+  type Fields = NamedTuple.Map[NamedTuple.From[T], Q]
+  def selectDynamic(fieldName: String) = ...
+```
+
+Assume in the user domain:
+```scala
+case class City(zipCode: Int, name: String, population: Int)
+val city: Q[City]
+```
+Then
+```scala
+city.zipCode
+```
+has type `Q[Int]` and it expands to
+```scala
+city.selectDynamic("zipCode").asInstanceOf[Q[Int]]
+```
+
+### The NamedTuple.From Type
+
+The `NamedTuple` object contains a type definition
+```scala
+  type From[T] <: AnyNamedTuple
+```
+`From` is treated specially by the compiler. When `NamedTuple.From` is applied to
+an argument type that is an instance of a case class, the type expands to the named
+tuple consisting of all the fields of that case class.
+Here, _fields_ means: elements of the first parameter section. For instance, assuming
+```scala
+case class City(zip: Int, name: String, population: Int)
+```
+then `NamedTuple.From[City]` is the named tuple
+```scala
+(zip: Int, name: String, population: Int)
+```
+The same works for enum cases expanding to case classes, abstract types with case classes as upper bound, alias types expanding to case classes
+and singleton types with case classes as underlying type (in terms of the implementation, the `classSymbol` of a type must be a case class).
+
+`From` is also defined on named tuples. If `NT` is a named tuple type, then `From[NT] = NT`.
+
+### Pattern Matching with Named Fields in General
+
+We allow named patterns not just for named tuples but also for case classes. For instance:
+```scala
+city match
+  case c @ City(name = "London") => println(p.population)
+  case City(name = n, zip = 1026, population = pop) => println(pop)
```

Comment:
Should we require a trailing `*` e.g. `case City(name = "London", *) =>` to make it clear that we are only matching a subset of fields?

Is there some way to do named-pattern-matching against _zero_ fields? This is different from `case c: City` because it's `unapply` based rather than `isInstanceOf`-based, and is more convenient than writing `case City(_, _, _) =>` which is the status quo

Reply:
> Is there some way to do named-pattern-matching against _zero_ fields? This is different from `case c: City` because it's `unapply` based rather than `isInstanceOf`-based, and is more convenient than writing `case City(_, _, _) =>` which is the status quo

I also think that `case City(*)` and would opt for that kind of syntax almost always than `_ : City`. This can be a different SIP, unless we require that indeed we need a trailing `*` if not all fields are specified.

Reply:
I am not fond of explicit * here. The situation for me is analogous to named function arguments. We don't use * there either to make clear that we have given all names and it's OK to use defaults for the rest. For pattern matching, the default is simply `_`, i.e. wildcard and no binding. 

Also, I think almost all named pattern matches will be partial. That's the whole point. If I have a full match for e.g. `Person`, then I would likely write:
```scala
case Person(name, age)
```
and not
```scala
case Person(name = name, age = age)
```
So in that sense, the fact that it is a named pattern match also indicates that it is partial, and the `*` would just be annoying and redundant.